### ISSUE TITLE
Hardinfo engine/subsystem idea

### ISSUE THREAD

**Issue by bp0** [lpereira/hardinfo#309]
I've had this idea for a while to make a sysfs browser that kindof interprets the items for you. I've finally gotten to play with it and I think it would make a good engine for hardinfo.

The idea is a tree of `sysobj` that are based around the items in the sysfs tree, where most of the data is coming from anyway.

The example screenshots below, you can see that `/sys/devices/system/cpu/cpu3` is made of parts of it's `cpufreq` and `topology` children. You can also see that `cache` is made up of it's `indexN` children, themselves made up of information from their own children `type`, `level`, `size`, etc.

Also, you can see a virtual sysobj called ":cpuinfo" which is made by parsing the `/proc/cpuinfo`, but each item in the tree can be queried in the same way as a regular sysfs item would. `:cpuinfo/logical_cpu3` is shown and it's flags in `:cpuinfo/logical_cpu3/flags`. Notice that a symlink to the sysfs cpu item for cpu3 is included as `:cpuinfo/logical_cpu3/_sysfs`

`:computer/board` works the same way as hardinfo's "method"  `devices::getMotherboard` works, using dmi or devicetree, whatever is available.

`/sys/bus/pci/devices` shows how a listing of PCI devices can automatically pull information from child items, in this case device's vendor name from pci.ids.

Please see https://github.com/bp0/verbose-spork to play with it.

![screenshot_2018-11-09_15-26-09](https://user-images.githubusercontent.com/1758090/48289728-1c5ca280-e436-11e8-9dc6-0ec1abf0da10.png)
![screenshot_2018-11-09_15-26-53](https://user-images.githubusercontent.com/1758090/48289729-1c5ca280-e436-11e8-881e-338102ca377c.png)
![screenshot_2018-11-09_15-27-15](https://user-images.githubusercontent.com/1758090/48289730-1c5ca280-e436-11e8-8607-e9d97c663d4e.png)
![screenshot_2018-11-09_15-28-00](https://user-images.githubusercontent.com/1758090/48289731-1c5ca280-e436-11e8-82f1-64f0dce030e0.png)
![screenshot_2018-11-09_15-28-12](https://user-images.githubusercontent.com/1758090/48289732-1c5ca280-e436-11e8-986a-0bbf7275cdb9.png)
![screenshot_2018-11-09_15-30-33](https://user-images.githubusercontent.com/1758090/48289733-1c5ca280-e436-11e8-8726-f6b115d7a83e.png)
![screenshot_2018-11-09_15-30-57](https://user-images.githubusercontent.com/1758090/48289734-1cf53900-e436-11e8-88b2-276b6f3a3d2b.png)
![screenshot_2018-11-09_15-31-33](https://user-images.githubusercontent.com/1758090/48289736-1cf53900-e436-11e8-8582-50b393d5fde7.png)

**Comment by lpereira** [comment]
Agreed, having a nice abstraction layer like this would improve things a lot in HardInfo.  Especially if we can figure out a way to use the struct Info stuff, so that most of it is automatic and not manual/haphazard.

**Comment by bp0** [comment]
@blacksvk , thanks. Yes, that is the idea. 

@lpereira , I think this makes the info struct stuff much easier because minimal/no formatting has to be done when the info struct is set up, it can be something like `info_field_from_sysobj("/sys/path/to/whatever", fmt_opts)`, and both the label and the formatted value may be supplied.

Mostly, I like the ability to snapshot a system and easily test against it. Also, there is a "pins" interface that monitors a list of sysobj's, refreshing at intervals that are reasonable for the whatever the item is. This can be seen in the "Watchlist" tab. The idea is you could "pin" anything, sensors or cpufreq or memory usage, or anything else in the /sys or virtual trees, perhaps during a benchmark or something. The "pin" can also keep a history and the min/max value.

**Comment by bp0** [comment]
Updated screenshots show some hwmon sensor formatting, gpu enumeration, using vendor.ids, and the last one shows the start of integration to Hardinfo.

![screenshot_2018-12-18_22-14-40](https://user-images.githubusercontent.com/1758090/50199827-ac84e600-0318-11e9-8cd2-2c4b2f3bfb5e.png)
![screenshot_2018-12-18_22-15-03](https://user-images.githubusercontent.com/1758090/50199841-bdcdf280-0318-11e9-8a79-2fdfdd8ef9de.png)
![screenshot_2018-12-18_22-23-39](https://user-images.githubusercontent.com/1758090/50199850-c6bec400-0318-11e9-8029-c2aca0262609.png)
![screenshot_2018-12-18_22-53-38](https://user-images.githubusercontent.com/1758090/50199862-d3431c80-0318-11e9-8389-bba7d0ae88f3.png)
![screenshot_2018-12-18_22-17-13](https://user-images.githubusercontent.com/1758090/50199902-0e455000-0319-11e9-9dfc-c5a6d01d484a.png)
![screenshot_2018-12-18_22-55-26](https://user-images.githubusercontent.com/1758090/50199908-130a0400-0319-11e9-8283-8d4e80d22fae.png)

**Comment by bp0** [comment]
One more:
![screenshot_2018-12-18_23-11-19](https://user-images.githubusercontent.com/1758090/50200240-a3951400-031a-11e9-80f9-4897f00b65db.png)

**Comment by lpereira** [comment]
This looks very cool! Good job!

**Comment by bp0** [comment]
As far as I can tell, the cell renderer for text does not support links in markup.

**Comment by lpereira** [comment]
It doesn't, but it does let you know the location of a click.
